### DISCUSSION TITLE
Does there exist a way to determine if an IDE plugin exists and/or is enabled?

### DISCUSSION THREAD

**Discussion by MostHated** [JetBrains/ideavim#808] [Q&A]
Hey there, I was wondering if there exist a way to determine if an IDE plugin exists and/or is enabled?

Reason being, sometimes I switch between Copilot and JB AI completions and I would like to be able to detect which one is enabled/active via conditional so that my keybind can call the right action.

Ex. A way to accomplish something like this:

```vim
if has(copilot.plugin) && isEnabled(copilot.plugin)
  imap <A-\> <Action>copilot.requestCompletions
else
  imap <A-\> <Action>AIAssistant.Editor.CompleteCodeWithAI
endif
```
Thanks,
-MH

**Comment by lippfi** [comment]
Hi! Nope, there is no such functionality currently. If you decide to create a PR to add the isPluginEnabled() function, you are very welcome to do so; it's a great idea.

P.S. It's not a good idea to have an if statement with two mappings inside, because the if in your config will be executed only once (during IDE startup). If you disable Copilot after the IDE starts, the mapping will not be aware of it. A better approach might be something like:
```
imap <expr> <A-\> isPluginEnabled(copilotId) ? "\<Action>(copilot.requestCompletions)" : "\<Action>(AIAssistant.Editor.CompleteCodeWithAI)"
```

**Comment by MostHated** [comment]
I appreciate the reply. I did realize that after I made my post while messing around with it some more.

What I was attempting to do a while later was:

```sh
# -- jb_plugin_check.sh

#!/bin/bash
PLUGIN_NAME="com.github.copilot"

is_plugin_disabled=$(grep -c "^$PLUGIN_NAME\$" ~/.config/JetBrains/Rider2024.1/disabled_plugins.txt)

if [[ $is_plugin_disabled -eq 1  ]]; then
    echo 1
else
    echo 0
fi
```

```vim
function! CoPilotDisabled()
  let check_script="/home/mosthated/_dev/languages/sh/apps/jetbrains/jb_plugin_check.sh"
  let result=system(check_script)

  if result == 0
    :action copilot.requestCompletions<cr>
  else
    :action AIAssistant.Editor.CompleteCodeWithAI<cr>
  endif
endfunction

imap <A-\> :call CoPilotDisabled()<CR> 
```

I was not able to get that working properly yet before I decided to just continue on with my work. I know I did something wrong, just wasn't sure yet what it was but after seeing your example, I may have another go at it later/tomorrow.

**Reply by lippfi** [reply]
Wow, it is clever. It won't work because the `system()` function is not supported by IdeaVim.
Our VimScript support is very basic; we've supported most of the language itself, but not its functions (there are about 600 of them if I'm not mistaken, and that's a lot of work). Currently, VimScript support is not as requested as other features, so the maintainers are focused on other issues.

However, contributing to VimScript functions is welcome (and easy; see [sin() function](https://github.com/JetBrains/ideavim/blob/master/vim-engine/src/main/kotlin/com/maddyhome/idea/vim/vimscript/model/functions/handlers/SinFunctionHandler.kt))

**Comment by MostHated** [comment]
I was honestly going to give it a go, but I don't use java often and have never used kotlin or gradle, and as such I have spent the last hour or two trying to install sdkman, get it working with intellij, but can't get past either this:

```sh
A problem occurred configuring root project 'IdeaVIM'.
> Failed to query the value of task ':runIde' property 'ideDir'.
   > Could not resolve all dependencies for configuration ':z90_intellij'.
      > Failed to calculate the value of task ':compileTestJava' property 'javaCompiler'.
         > No matching toolchains found for requested specification: {languageVersion=17, vendor=any, implementation=vendor-specific} for LINUX on x86_64.
            > No locally installed toolchains match and toolchain download repositories have not been configured.

* Try:
> Learn more about toolchain auto-detection at https://docs.gradle.org/8.5/userguide/toolchains.html#sec:auto_detection.
> Learn more about toolchain repositories at https://docs.gradle.org/8.5/userguide/toolchains.html#sub:download_repositories.
> Run with --stacktrace option to get the stack trace.
> Run with --info or --debug option to get more log output.
> Run with --scan to get full insights.
> Get more help at https://help.gradle.org.
```
or it telling me I am not using a supported version.

I am just frustrated now, and going to continue on with my day.

**Reply by lippfi** [reply]
You should navigate to the Project Structure, download any Java 17 SDK, and this will resolve the issue
![image](https://github.com/JetBrains/ideavim/assets/81118900/d58c66db-30ae-449d-8bdf-cbeedfaf0729)

**Comment by MostHated** [comment]
I see. That did seem to work. I checked all over the contributors document, and I don't recall seeing a version of 17 listed. How is one supposed to know? 

I realize that PR's are a good thing, but I don't think you would want them from someone who knows nothing about the language or it's best practices, let alone how to even set up the ide.

That said, I may have a go at it tomorrow, after I look over a bunch of additional plugin examples and see if I can familiarize myself on a smaller scale first.

**Reply by lippfi** [reply]
It's listed in [gradle.properties](https://github.com/JetBrains/ideavim/blob/f4afdb21b2490ee50b54047df90c44cf7407574d/gradle.properties#L15), but trying to run a project and gathering information from an exception is also a valid approach.

We will review your request before merging, so don't worry about this; everyone does things for the first time.

Let me give you some tips that may be helpful:


1. IdeaVim is divided into two folders: src and vim-engine. The vim-engine contains common Vim logic and will become a Vim library for different IDEs in the future, and it does not depend on the IntelliJ Platform. src contains code that depends on the IntelliJ Platform.
Because of this, we have two places for storing functions: [IntelliJ Platform dependent](https://github.com/JetBrains/ideavim/tree/master/src/main/java/com/maddyhome/idea/vim/vimscript/model/functions/handlers) and [IntelliJ Platform independent](https://github.com/JetBrains/ideavim/tree/master/vim-engine/src/main/kotlin/com/maddyhome/idea/vim/vimscript/model/functions/handlers).
How is this related to you: If you decide to implement `system()`, it should be in the vim-engine, since you do not need any dependencies on the IntelliJ Platform, and it is also something common that feels right in the Vim core. However, if you decide to implement `isPluginEnabled()`, you should do it in the src folder, because it will be helpful only for Platform users and also requires some Platform-dependent code.
2. Most likely, the IntelliJ Platform's `com.intellij.ide.plugins.PluginManagerCore` class is what you are looking for to implement the isPluginEnabled() function.
3. ChatGPT knows a lot about the IntelliJ Platform and can suggest some useful classes or give tips (it was ChatGPT who told me about `com.intellij.ide.plugins.PluginManagerCore`; I had never heard about this class before).

Good luck, and feel free to ask questions if there are any.
